File: src/main/java/maple/expectation/global/executor/SafeExecutor.java
package maple.expectation.global.executor;

import maple.expectation.global.common.function.ThrowingSupplier;

/**
 * Safe execution pattern interface.
 *
 * <p>Provides fallback capabilities for error recovery. Part of Interface Segregation
 * Principle (ISP) compliance - clients only depend on methods they use.
 */
public interface SafeExecutor {

  /**
   * Execute task with default value on exception.
   *
   * @param <T> result type
   * @param task task to execute
   * @param defaultValue default value on exception
   * @param context execution context
   * @return task result or default
   */
  <T> T executeOrDefault(ThrowingSupplier<T> task, T defaultValue, TaskContext context);
}

File: module-infra/src/main/kotlin/maple/expectation/infrastructure/executor/SafeExecutor.kt
package maple.expectation.infrastructure.executor

import maple.expectation.common.function.ThrowingSupplier

/**
 * Safe execution pattern interface.
 *
 * Provides fallback capabilities for error recovery.
 * Part of Interface Segregation Principle (ISP) compliance - clients only depend on methods they use.
 */
interface SafeExecutor {
    /**
     * Execute task with default value on exception.
     */
    fun <T> executeOrDefault(task: ThrowingSupplier<T>, defaultValue: T, context: TaskContext): T
}
